// dtcarrow/section.js
// Library code specifically for section api code
var Arrow = require('arrow'),
    Q = require('q'),
    Dtcarrow = require('dtcarrow');

// CRUD
module.exports.create = function(section) {
    // // Writes a new section to the Section table
    var deferred = Q.defer();

    var check = {
        sectionTypeId: 'string',
        value: 'string'
    }
    var checkedSection = Dtcarrow.Common.checkObject(section, check);
    if (!checkedSection) {
        deferred.reject({ success: false, status: 422, caller: 'section.js>create',
            data: 'Inbound object does not match expectations' });
        return deferred.promise;
    }

    var Model = Arrow.getModel('appc.mysql.eolmig/Section');
    Model.create(checkedSection, function(err, results) {
        if (err) {
            deferred.reject({ success: false, status: 500, caller: 'section.js>create',
                data: err });
            return deferred.promise;
        }

        deferred.resolve({ success: true, status: 201, caller: 'section.js>create',
                data: results });
    });

    return deferred.promise;
};

module.exports.read = function(sectionId) {
    var deferred = Q.defer();

    if (typeof sectionId === 'undefined') {
        deferred.reject({ success: false, status: 422, caller: 'section.js>create',
            data: 'Missing sectionId' });
        return deferred.promise;
    }

    var Model = Arrow.getModel('appc.mysql.eolmig/Section');
    var query = 'SELECT * FROM Section WHERE sectionId = ' + sectionId + ' LIMIT 1';
    Model.query({ customSqlQuery: query }, function(err, results) {
        if (err) {
            deferred.reject({ success: false, status: 500, caller: 'section.js>read', 
                data: err });
            return deferred.promise;
        }

        if (typeof results === 'undefined' || results.length === 0) {
            deferred.reject({ success: false, status: 422, caller: 'section.js>read',
                data: 'No results' });
            return deferred.promise;
        }

        deferred.resolve({ success: true, status: 200, caller: 'section.js>read',
                data: results });
    });

    return deferred.promise;
};


module.exports.readAll = function() {
    // Gets a list of all sections via the Section table
    var deferred = Q.defer();

    var Model = Arrow.getModel('appc.mysql.eolmig/Section');
    var query = 'SELECT * FROM Section'
    Model.query({ customSqlQuery: query }, function(err, results) {
        if (err) {
            deferred.reject({ success: false, status: 500, caller: 'section.js>readAll',
                data: err });
            return deferred.promise;
        }

        if (typeof results === 'undefined' || results.length === 0) {
            deferred.reject({ success: false, status: 422, caller: 'section.js>readAll',
                data: 'No results' });
            return deferred.promise;
        }

        deferred.resolve({ success: true, status: 200, caller: 'section.js>readAll',
                data: results });
    });

    return deferred.promise;
};

module.exports.update = function(section) {
    // Updates existing section record
    var deferred = Q.defer();

    var check = {
        sectionId: 'string',
        description: 'string'
    }
    var checkedSection = Dtcarrow.Common.checkObject(section, check);
    if (!checkedSection) {
        deferred.reject({ success: false, status: 422, caller: 'section.js>update',
            data: 'Inbound object does not match expectations' });
        return deferred.promise;
    }

    var Model = Arrow.getModel('appc.mysql.eolmig/Section');
    var query = 'UPDATE Section SET description = "' + checkedSection.description +
              + ' WHERE sectionId = ' + checkedSection.sectionId;
    Model.query({ customSqlQuery: query }, function(err, results) {
        if (err) {
            deferred.reject({ success: false, status: 500, caller: 'section.js>update',
                data: err });
            return deferred.promise;
        }

        deferred.resolve({ success: true, status: 200, caller: 'section.js>update',
                data: section });
    });

    return deferred.promise;
};

module.exports.delete = function(section) {
    // // Deletes a record from the Section table
    var deferred = Q.defer();

    if (typeof section.sectionId === 'undefined') {
        deferred.reject({ success: false, status: 422, caller: 'section.js>delete',
            data: 'Missing sectionId' });
        return;   
    }

    var Model = Arrow.getModel('appc.mysql.eolmig/Section');
    var query = 'DELETE FROM Section WHERE sectionId = ' + section.sectionId;
    Model.query({ customSqlQuery: query }, function(err, rs) {
        if (err) {
            deferred.reject({ success: false, status: 500, caller: 'section.js>delete',
                data: err });
            return deferred.promise;
        }

        deferred.resolve({ success: true, status: 204, caller: 'section.js>delete',
                data: true });
    });

    return deferred.promise;
}
